internal/metrics/prometheus.go

```diff
@@ -0,0 +1,170 @@
+				KonnectServerURLKey,
+				KonnectEntityOperationTypeKey, KonnectEntityOperationCreate, KonnectEntityOperationUpdate, KonnectEntityOperationDelete,
+				KonnectEntityTypeKey,
+				SuccessKey, SuccessTrue, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+			),
+		},
+		[]string{KonnectServerURLKey, KonnectEntityOperationTypeKey, KonnectEntityTypeKey, SuccessKey, StatusCodeKey},
+	)
+
+	konnectEntityOperationDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: MetricNameKonnectEntityOperationDuration,
+			Help: fmt.Sprintf(
+				"How long did the Konnect entity operation take in milliseconds. "+
+					"`%s` describes the URL of the Konnect server. "+
+					"`%s` describes the operation type (`%s`, `%s`, or `%s`)."+
+					"`%s` describes the type of the operated entity. "+
+					"`%s` describes whether the operation is successful (`%s`) or not (`%s`). "+
+					"`%s` is populated in case of `%s=\"%s\"` and describes the status code returned from Konnect API. "+
+					"`%s`=\"0\" and %s=\"%s\" means we cannot collect the status code or error happens in the process other than Konnect API call.",
+				KonnectServerURLKey,
+				KonnectEntityOperationTypeKey, KonnectEntityOperationCreate, KonnectEntityOperationUpdate, KonnectEntityOperationDelete,
+				KonnectEntityTypeKey,
+				SuccessKey, SuccessTrue, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+			),
+			// Duration range from 1ms to 10min.
+			Buckets: prometheus.ExponentialBucketsRange(1, 10*float64(time.Minute.Milliseconds()), 20),
+		},
+		[]string{KonnectServerURLKey, KonnectEntityOperationTypeKey, KonnectEntityTypeKey, SuccessKey, StatusCodeKey},
+	)
+)
+
+// GlobalCtrlRuntimeMetricsRecorder is a metrics recorder that uses a global Prometheus registry
+// provided by the controller-runtime. Any instance of it will record metrics to the same registry.
+//
+// We want to expose Gateway operator's custom metrics on the same endpoint as controller-runtime's built-in
+// ones. Because of that, we have to use its global registry as CR doesn't allow injecting a custom one.
+// Upstream issue regarding this: https://github.com/kubernetes-sigs/controller-runtime/issues/210.
+type GlobalCtrlRuntimeMetricsRecorder struct{}
+
+var _ Recorder = &GlobalCtrlRuntimeMetricsRecorder{}
+
+func NewGlobalCtrlRuntimeMetricsRecorder() *GlobalCtrlRuntimeMetricsRecorder {
+	return &GlobalCtrlRuntimeMetricsRecorder{}
+}
+
+func (r *GlobalCtrlRuntimeMetricsRecorder) RecordKonnectEntityOperationSuccess(
+	serverURL string, operationType KonnectEntityOperation, entityType string, duration time.Duration) {
+	r.recordKonnectEntityOperationCount(serverURL, operationType, entityType, true, 0)
+	r.recordKonnectEntityOperationDuration(serverURL, operationType, entityType, true, 0, duration)
```

Comment:
I believe the `0` will show up in the metric's label values here which might be confusing for users 🤔 

We do seem to not return the response status codes from Konnect ops, e.g. [`createSNI()`](https://github.com/Kong/gateway-operator/blob/8c7fbd40e945bb95c5247288131eac3df9cc7113/controller/konnect/ops/ops_kongsni.go#L21-L51).

I believe that if we do not want to refactor all of these functions to return the status code we should include the note in metrics help text stating that `0` is the expected value for succees. WDYT?

Reply:
Added the introduction in helps and comments.